src/Curation/Curation.service.ts

```diff
@@ -0,0 +1,116 @@
+import { raw, SQL } from 'decentraland-server'
+import { hasAccess as hasCollectionAccess } from '../Collection/access'
+import { getMergedCollection } from '../Collection/utils'
+import { hasAccess as hasItemAccess } from '../Item/access'
+import { getMergedItem } from '../Item/utils'
+import { CollectionCuration } from './CollectionCuration'
+import { CurationType } from './Curation.types'
+import { ItemCuration } from './ItemCuration'
+
+export class CurationService<
+  T extends typeof CollectionCuration | typeof ItemCuration
+> {
+  public type: CurationType
+
+  // TODO: Type this better (or remove it completely and use+mock new CurationService)
+  static byType(type: CurationType): CurationService<any> {
+    switch (type) {
+      case CurationType.COLLECTION:
+        return new CurationService(CollectionCuration)
+      case CurationType.ITEM:
+        return new CurationService(ItemCuration)
+      default:
+        throw new Error(`Invalid curation type ${type}`)
+    }
+  }
+
+  constructor(public CurationModel: T) {
+    switch (CurationModel) {
+      case CollectionCuration:
+        this.type = CurationType.COLLECTION
+        break
+      case ItemCuration:
+        this.type = CurationType.ITEM
+        break
+      default:
+        throw new Error(`Invalid curation model ${CurationModel}`)
+    }
+  }
+
+  async getLatest() {
```

Comment:
What's the type `getLatest` returns? We could use generics to say that this method will return either a CollectionCuration or an ItemCuration.

Reply:
Well crap, it seems that at some point along the development of the feature I broke the typings. These are not so trivial to type sadly, I'd have to do something like this:

```ts
// Either this:
async getLatest() {
  const columnName = this.getForeingColumnName()
  return this.getModel().query<T['prototype']['attributes']>( ... )
}

// or:
async getLatest(): Promise<T['prototype']['attributes'][]> {
  const columnName = this.getForeingColumnName()
  return this.getModel().query( ... )
}
```
----------------------

My first implementation revolved around the `CurationType` like this:

```ts
export class CurationService<T extends CurationType> {
  public CurationModel: T extends CurationType.COLLECTION
    ? typeof CollectionCuration
    : typeof ItemCuration

  constructor(public type: T) {
    switch (type as CurationType) {
      case CurationType.COLLECTION:
        this.CurationModel = CollectionCuration
        break
      case CurationType.ITEM:
        this.CurationModel = ItemCuration
        break
      default:
        throw new Error(`Invalid curation model ${type}`)
    }
  }
```

which I think looks cleaner, but I can't get Typescript to agree with me
![image](https://user-images.githubusercontent.com/692077/147990604-5c417bf2-4549-4595-bee4-a7068096fbfb.png)

Also, typing `getLatest` in this situation'd look something like

```ts
async getLatest() {
  const columnName = this.getForeingColumnName()
  return this.getModel().query<
    typeof this.CurationModel['prototype']['attributes']
  >(...) 
```

(I think)



Do you have a better idea? which one do you like best?